IEP-902 DFU toggle button is odd

sigmaaa opened the pull request:
## Description

DFU toggle button is odd, because it modifies the launch configuration in a non-obvious way. In the launch configuration we still see normal "idf.py flash" command, but when you click the launch button, DFU is done instead. The button can be enabled even for targets which don't support DFU at all, which produces a non-obvious error (https://github.com/espressif/esp-idf/issues/10656, and also one customer at the workshop also ran into this.) I would suggest moving this setting into the launch configuration. Or allowing the user to create different types of launch configs — flash over UART, flash over DFU, flash over JTAG.

Fixes # ([IEP-902](https://jira.espressif.com:8443/browse/IEP-902))

## Type of change

Please delete options that are not relevant.
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?

Test 1:
- open a launch configuration, and switch between flash-over options (DFU, UART, JTAG) and test if composite is updated properly (No UI changes for DFU for now)
Test 2:
- flash the project with the selected interface

**Test Configuration**:
* ESP-IDF Version:
* OS (Windows,Linux and macOS):

## Dependent components impacted by this PR:

- Build command
- Launch configuration

## Checklist
- [ ] PR Self Reviewed
- [ ] Applied Code formatting
- [ ] Added Documentation
- [ ] Added Unit Test
- [ ] Verified on all platforms - Windows,Linux and macOS

sigmaaa commented:
Hi @kolipakakondal, @alirana01

I want to add these changes https://github.com/espressif/idf-eclipse-plugin/pull/318 here, but having some difficulties related to disposing of composites, which caused some bugs. 
After fixing them, I'll update UI for the DFU option. I'm planning to add a clean composite with one combo with filtered targets esp32s2 and esp32s3

kolipakakondal commented:
Hi @sigmaaa Thanks for the PR. Changes LGTM. We could address the following issues as well.

1. When DFU is selected, UI is not reflected. We can update the arguments section with the `idf.py dfu-flash` command(I think you already mentioned this.) Not sure if dfu-flash accepts any other arguments. If not, we can get rid of complete UI.
2. DFU Icon(disabled state) still visible on toolbar section
3. Update DFU usage documentation 
4. Update existing JTAG Flashing documentation with the new screenshots
5. I see the below popup even when DFU is selected, it could be due to JTAG UI still showing when DFU is selected.


<img width="1440" alt="Screenshot 2023-03-16 at 5 45 22 AM" src="https://user-images.githubusercontent.com/8463287/225488963-e31a3035-3e7d-4a47-9bc7-5bf52dab5a86.png">

kolipakakondal commented:
Hi @sigmaaa Thanks for the updates. I could see some UI issues. Please check below.

Need to fix the spelling
<img width="548" alt="Screenshot 2023-03-20 at 9 48 50 AM" src="https://user-images.githubusercontent.com/8463287/226246337-5f2eab71-4954-4c47-985e-5bdde5dded96.png">

An issue with rending when I switch between different flash options
<img width="957" alt="Screenshot 2023-03-20 at 9 48 19 AM" src="https://user-images.githubusercontent.com/8463287/226246373-eb2f24c3-7691-4e47-bf80-6ac8ff0b4bf0.png">

kolipakakondal commented:
<img width="963" alt="Screenshot 2023-03-20 at 9 53 45 AM" src="https://user-images.githubusercontent.com/8463287/226246701-ba674f0c-a435-42a7-9169-39102612844a.png">

Maybe we can rephrase the error - 
 The active launch target esp32 does not support DFU. The DFU is supported only by esp32s2 and esp32s3

kolipakakondal commented:
<img width="911" alt="Screenshot 2023-03-20 at 10 02 03 AM" src="https://user-images.githubusercontent.com/8463287/226247586-3e6de399-89f2-47f9-aa1e-2468ff2f4995.png">

I think this is a wrong error message popup - it was triggered when I switch from JTAG to the DFU

kolipakakondal commented:
<img width="486" alt="Screenshot 2023-03-20 at 10 04 11 AM" src="https://user-images.githubusercontent.com/8463287/226247963-3c82aa19-db4a-4b68-8f80-79d23dbba293.png">

Can we move the `Targets with DFU Support' drop-down right below the Flash Over drop-down when DFU is selected? This will create uniformity in the UI

alirana01 commented on bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:
```diff
@@ -74,48 +77,229 @@
+		Label comboTargetLbl = new Label(targetComposite, SWT.NONE);
+		comboTargetLbl.setText(Messages.CMakeMainTab2_TargetsComboLbl);
+		comboTargets = new Combo(targetComposite, SWT.DROP_DOWN | SWT.READ_ONLY);
+		comboTargets.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, false, false));
```
Can we try to avoid SWT.CENTER? This has previously caused issues when rendering SWT controls on Linux. I faced the issue with Install Tools Wizard.

sigmaaa replied:
Hi @alirana01, thank you for reviewing. Fixed

sigmaaa commented:
> <img alt="Screenshot 2023-03-20 at 10 02 03 AM" width="911" src="https://user-images.githubusercontent.com/8463287/226247586-3e6de399-89f2-47f9-aa1e-2468ff2f4995.png">
> 
> I think this is a wrong error message popup - it was triggered when I switch from JTAG to the DFU

Hi @kolipakakondal, thank you for reviewing. Fixed by adding an `ignore` flag, which is activated in LaunchBarListener class while we edit the launch configuration. 

also fixed the issue with spelling and changed the error message.

sigmaaa commented:
> <img alt="Screenshot 2023-03-20 at 10 04 11 AM" width="486" src="https://user-images.githubusercontent.com/8463287/226247963-3c82aa19-db4a-4b68-8f80-79d23dbba293.png">
> 
> Can we move the `Targets with DFU Support' drop-down right below the Flash Over drop-down when DFU is selected? This will create uniformity in the UI

Fixed here https://github.com/espressif/idf-eclipse-plugin/pull/716/commits/0f0ca37254ca5fa3d8d7b773a6a69888fa06bd68

sigmaaa commented:
Hi @kolipakakondal, 

can't reproduce this behavior on mac and windows:
![image](https://user-images.githubusercontent.com/24419842/226348219-3b650a38-543b-4dd8-bac3-6ba85143e9b7.png)

kolipakakondal reviewed (approved):
LGTM

AndriiFilippov commented:
hi @sigmaaa ! 

Tested under: 
OS - Windows 10  / Linux Ubintu
ESP-IDF: v5.0 

LGTM 👍
